File: 20220817/모범답안/20220817_06.js/CountBinaryCard.js
/**
 * @param n {number}
 * @return {number}
 */
function solution(n) {
  // 중복된 결과를 저장
  const dp = [0];

  // 2의 배수로 반복되는 bit 처리를 위함
  let offset = 1;

  for (let i = 1; i <= n; i++) {
    if (offset * 2 === i) offset *= 2;

    dp[i] = dp[i - offset] + 1;
  }

  // 0부터 n까지 필요한 붉은색 종이수
  return dp.reduce((a, b) => a + b, 0);
}

solution

File: 20220817/모범답안/20220817_06.js/CountBinaryCard.py
from functools import reduce


def solution(n):
    """
    :param n: int
    :return: int
    """

    # 중복된 결과를 저장
    dp = [0]

    # 2의 배수로 반복되는 bit 처리를 위함
    offset = 1

    for i in range(1, n + 1):
        if offset * 2 == i:
            offset *= 2

        dp.insert(i, dp[i - offset] + 1)

    # 0부터 n까지 필요한 붉은색 종이수
    return reduce(lambda a, b: a + b, dp, 0)
